ncd-calculator/src/services/kgrid.ts

```diff
@@ -0,0 +1,328 @@
+            const factor = gradualFactor({i, j});
+
+            total += ncdValue * factor;
+        }
+    }
+
+    // vertical neighbors
+    for(let i = 0; i < grid.height; i++) {
+        for(let j = 0; j < grid.width; j++) {
+            const down = (i + 1) % grid.height;
+
+            const obj1 = grid.objects[i][j];
+            const obj2 = grid.objects[down][j];
+
+            const ncdFactor = grid.ncdMatrix[obj1.id][obj2.id];
+            const factor = gradualFactor({i, j});
+
+            total += ncdFactor * factor;
+        }
+    }
+
+    return total;
+}
+
+export const deepCopy = (grid: GridState): GridState => {
+    const newGrid: GridState = {
+        height: grid.height,
+        width: grid.width,
+        objectiveValue: grid.objectiveValue,
+        ncdMatrix: [],
+        objects: [],
+    };
+
+    newGrid.ncdMatrix = grid.ncdMatrix.map(row => [...row]);
+
+    newGrid.objects = [];
+    for(let i = 0; i < grid.height; i++) {
+        newGrid.objects[i] = [];
+        for(let j = 0; j < grid.width; j++) {
+            newGrid.objects[i][j] = {
+                id: grid.objects[i][j].id,
+                content: grid.objects[i][j].content
+            }
+        }
+    }
+
+    return newGrid;
+}
+
+
+export const swapBlocks = (
+    grid: GridState,
+    block1: Block,
+    block2: Block,
+    reflection: [boolean, boolean]
+): GridState => {
+    const newGrid = deepCopy(grid);
+
+    const block1Content = extractBlock(grid, block1);
+    const block2Content = extractBlock(grid, block2);
+
+
+    if (reflection[0]) {
+        reflectHorizontally(block1Content);
+        reflectHorizontally(block2Content)
+    }
+    if (reflection[1]) {
+        reflectVertically(block1Content);
+        reflectVertically(block2Content);
+    }
+
+    placeBlock(newGrid, block2, block1Content);
+    placeBlock(newGrid, block1, block2Content);
+
+    newGrid.objectiveValue = calculateObjective(newGrid);
+    return newGrid;
+}
+
+export const extractBlock = (grid: GridState, block: Block): GridObject[][] => {
+    const {topLeft, bottomRight} = block;
+
+    const blockHeight = bottomRight.i - topLeft.i + 1;
+    const blockWidth = bottomRight.j - topLeft.j + 1;
+
+    const blockContent: GridObject[][] = [];
+
+    for(let i = 0; i < blockHeight; i++) {
+        blockContent[i] = [];
+        for(let j = 0; j < blockWidth; j++) {
+            const gridI = topLeft.i + i;
+            const gridJ = topLeft.j + j;
+
+            if (gridI >= 0 && gridI < grid.height && gridJ >= 0 && gridJ < grid.width) {
+                blockContent[i][j] = {
+                    id: grid.objects[gridI][gridJ].id,
+                    content: grid.objects[gridI][gridJ].content
+                };
+            } else {
+                // Handle out-of-bounds gracefully
+                console.warn(`Block coordinates out of bounds: (${gridI},${gridJ})`);
+                blockContent[i][j] = {
```

Comment:
i don't think we need to worry about out of bounds coordinates. the computer can always generate positions that are in the grid bounds.